packages/dev/sharedUiComponents/src/historyStack.ts

```diff
@@ -162,17 +153,18 @@ export class HistoryStack implements IDisposable {
      * Undo the latest operation
      */
     public undo() {
-        if (this._history.length < 2) {
+        if (!this._historyStack.length) {
             return;
         }
 
         this._locked = true;
-        const current = this._history.pop()!;
-        this._redoStack.push(current);
+        const diff = this._historyStack.pop()!;
 
-        const newState = this._rebuildState();
+        const newState = this._applyJSONDiff(this._activeData, JSON.parse(diff));
+        this._redoStack.push(JSON.stringify(this._generateJSONDiff(newState, this._activeData)));
 
-        this._applyUpdate(newState);
+        this._applyUpdate(this._copy(newState));

```

Comment:
I don't think you have to call `_copy` here? `newState` is not stored, so even if the receiver of `newState` updates the object, it won't have any bad side-effect(?)

Reply:
This is actually the same response for your 3 questions: if I do not store a copy, the editor will update certain properties on the active graph (like locations) and that will change the diff of the active data. It took me a while to figure that one out:)